diff/windows/windows.go

```diff
@@ -191,3 +326,54 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 
 	return mnt.Source, parentLayerPaths, nil
 }
+
+// Ensure the two sets of mount-lists are actually a correct parent-and-child, or
+// orphan-and-empty-list, and return the full list of layers, starting with the
+// upper-most (most childish?)
+func mountPairToLayerStack(lower, upper []mount.Mount) ([]string, error) {
+
+	// May return an ErrNotImplemented, which will fall back to LCOW
+	upperLayer, upperParentLayerPaths, err := mountsToLayerAndParents(upper)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Upper mount invalid")
+	}
+
+	// Trivial case, diff-against-nothing
+	if len(lower) == 0 {
+		if len(upperParentLayerPaths) != 0 {
+			return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer with parents against a null layer")
+		}
+		return []string{upperLayer}, nil
+	}
+
+	if len(upperParentLayerPaths) < 1 {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer with no parents against another layer")
+	}
+
+	lowerLayer, lowerParentLayerPaths, err := mountsToLayerAndParents(lower)
+	if errdefs.IsNotImplemented(err) {
+		// Upper was a windows-layer, lower is not. We can't handle that.
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a windows-layer against a non-windows-layer")
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "Lower mount invalid")
+	}
+
+	if upperParentLayerPaths[0] != lowerLayer || len(upperParentLayerPaths) != len(lowerParentLayerPaths)+1 {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer against a layer other than its own parent")
+	}
+	for i, upperParent := range upperParentLayerPaths[1:] {
+		if upperParent != lowerParentLayerPaths[i] {
+			return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer against a layer with different parents")
+		}
+	}
+
+	return append([]string{upperLayer}, upperParentLayerPaths...), nil
+}
+
+func uniqueRef() string {
+	t := time.Now()
+	var b [3]byte
+	// Ignore read failures, just decreases uniqueness
+	rand.Read(b[:])
+	return fmt.Sprintf("%d-%s", t.UnixNano(), base64.URLEncoding.EncodeToString(b[:]))
+}
```

Comment:
This looks like the same function as defined in diff/walking/differ.go maybe we should move it to a shared utils file?

Reply:
I indeed just copied it from there. That's related to [what I wrote earlier](https://github.com/containerd/containerd/pull/4399#issuecomment-660283335): the `Compare` function it's called from is also basically the same as the one in the naive differ, and they should be refactored to share the common parts. I copied this one as-is because I wasn't sure we'd want to make this a public API and (at the time) I was unaware of things like the `internal/` package tree.

I think this function also appears in a third place in the codebase, too.

Reply:
I've identified five copies of this function (including the new one I've added here), plus one semantic-equivalent:
* unpacker.go `uniquePart` (uses `t.Nanosecond()`)
* diff/walking/differ.go `uniqueRef` (uses `t.UnixNano()`) -- This is the one I copied, because `(s windowsDiff) Compare` itself started life as a copy of `(s *walkingDiff) Compare`.
* diff/windows/windows.go `uniqueRef` (uses `t.UnixNano()`) -- The new one I'm adding here.
* leases/id.go `WithRandomID`  (uses `t.Nanosecond()`)
* pkg/cri/util/id.go `GenerateID` (no timestamp, uses `hex.EncodeToString` instead of `base64.URLEncoding.EncodeToString`)
* rootfs/apply.go `uniquePart` (uses `t.Nanosecond()`)

I honestly don't see a good place to put a unified copy of this, structure-wise. I don't see a util or helper package, and there isn't an _/internal_ tree at this time to add module-local helpers to.